### ISSUE TITLE
RTL8821ae drops wifi with kernels 4.10.0-27 and -28 and Ubuntu 16.04.2 LTS

### ISSUE THREAD

**Issue by NickBT** [lwfinger/rtlwifi_new#261]
I have a laptop running Ubuntu 16.02-2 LTS with hardware enablement stack. With kernel 4.8.-58 the wifi was fine. An updated prompted by Ubuntu took the kernel to a 4.10 version. After the update the wifi drops if a lot of data is transmitted to my laptop  (e.g BBC's iplayer or a speedtest site is used). After the wifi drops, the command **sudo service network-manager restart** does not turn on the wifi again.

The wifi configuration as produced by a **sudo lshw -C network** is:

 *-network
       description: Wireless interface
       product: RTL8821AE 802.11ac PCIe Wireless Network Adapter
       vendor: Realtek Semiconductor Co., Ltd.
       physical id: 0
       bus info: pci@0000:03:00.0
       logical name: wlp3s0
       version: 00
       serial: b0:c0:90:3b:d8:f3
       width: 64 bits
       clock: 33MHz
       capabilities: pm msi pciexpress bus_master cap_list ethernet physical wireless
       configuration: broadcast=yes driver=rtl8821ae driverversion=4.10.0-27-generic firmware=N/A ip=192.168.0.12 latency=0 link=yes multicast=yes wireless=IEEE 802.11
       resources: irq:51 ioport:3000(size=256) memory:b2000000-b2003fff

Fortunately I have a GRUB option to start with the retained 4.8 kernel but I would like to be able to use the latest 4.10 version.

**Comment by NickBT** [comment]
The same error is now exhibited with the newly released kernel version 4.10.0-30.
I was hoping that my post might draw some response from lwfinger.

**Comment by jhod0** [comment]
I am having this error with Kubuntu 17.04.

```
$ uname -r
4.10.0-32-generic
```

**Comment by AryToNeX** [comment]
Did someone solve this problem? I'm facing this too.

**Comment by lwfinger** [comment]
A patch is in the pipeline. See https://patchwork.kernel.org/patch/9962615/. As the problem only occurred in kernel 4.11 and later for me, I have no idea if it will fix your problem, or not. Ubuntu has this really strange way of patching kernels, thus it might.

**Comment by AryToNeX** [comment]
I'm on Ubuntu 16.04 LTS, kernel 4.10.0 with rtlwifi_new drivers and still my wireless disconnects and becomes unresponsive randomly, no matter what WiFi connection it is

**Comment by lwfinger** [comment]
Switch to branch "kernel_version" and try it.

**Comment by AryToNeX** [comment]
```
make -C /lib/modules/4.10.0-35-generic/build M=/home/anthony/rtlwifi_new modules
make[1]: ingresso nella directory "/usr/src/linux-headers-4.10.0-35-generic"
  CC [M]  /home/anthony/rtlwifi_new/base.o
/home/anthony/rtlwifi_new/base.c: In function ‘rtl_make_smps_action’:
/home/anthony/rtlwifi_new/base.c:2138:17: error: implicit declaration of function ‘skb_put_zero’ [-Werror=implicit-function-declaration]
  action_frame = skb_put_zero(skb, 27);
                 ^
/home/anthony/rtlwifi_new/base.c:2138:15: warning: assignment makes pointer from integer without a cast [-Wint-conversion]
  action_frame = skb_put_zero(skb, 27);
               ^
/home/anthony/rtlwifi_new/base.c: In function ‘rtl_make_del_ba’:
/home/anthony/rtlwifi_new/base.c:2267:15: warning: assignment makes pointer from integer without a cast [-Wint-conversion]
  action_frame = skb_put_zero(skb, 34);
               ^
cc1: some warnings being treated as errors
scripts/Makefile.build:294: set di istruzioni per l'obiettivo "/home/anthony/rtlwifi_new/base.o" non riuscito
make[2]: *** [/home/anthony/rtlwifi_new/base.o] Errore 1
Makefile:1524: set di istruzioni per l'obiettivo "_module_/home/anthony/rtlwifi_new" non riuscito
make[1]: *** [_module_/home/anthony/rtlwifi_new] Errore 2
make[1]: uscita dalla directory "/usr/src/linux-headers-4.10.0-35-generic"
Makefile:47: set di istruzioni per l'obiettivo "all" non riuscito
make: *** [all] Errore 2
```

**Comment by mirh** [comment]
kernel_version is based on 4.13

**Comment by AryToNeX** [comment]
Can you regularly get kernel 4.13 on Xubuntu 16.04 LTS? Nope, because 17.10 isn't out, so no 16.04.4 based on 4.13

I tried 17.10 beta 2 based on 4.13 and the issue is still there btw

**Comment by lwfinger** [comment]
I fixed the error reported by @AryToNeX. The code will now build for kernels 4.7 or newer.

@mirh - Actually, it is based on  kernel 4.14 from wireless-drivers-next! The fix for the issue is pending for mainline 4.14 and will be backported to 4.13, 4.12, and 4.11 once it makes it to mainline.

**Comment by DozyDolphin** [comment]
I'm on a desktop running Ubuntu 16.04.3, with kernel 4.10.0-35. I have a rtl8812ae chipset.

I just tried switching to the kernel_version branch and build with no luck. I'm getting the same errors as AryToNex are reporting:

```
make -C /lib/modules/4.10.0-35-generic/build M=/opt/rtlwifi/rtlwifi_new modules
make[1]: Entering directory '/usr/src/linux-headers-4.10.0-35-generic'
  CC [M]  /opt/rtlwifi/rtlwifi_new/base.o
/opt/rtlwifi/rtlwifi_new/base.c: In function ‘rtl_make_smps_action’:
/opt/rtlwifi/rtlwifi_new/base.c:2138:17: error: implicit declaration of function ‘skb_put_zero’ [-Werror=implicit-function-declaration]
  action_frame = skb_put_zero(skb, 27);
                 ^
/opt/rtlwifi/rtlwifi_new/base.c:2138:15: warning: assignment makes pointer from integer without a cast [-Wint-conversion]
  action_frame = skb_put_zero(skb, 27);
               ^
/opt/rtlwifi/rtlwifi_new/base.c: In function ‘rtl_make_del_ba’:
/opt/rtlwifi/rtlwifi_new/base.c:2267:15: warning: assignment makes pointer from integer without a cast [-Wint-conversion]
  action_frame = skb_put_zero(skb, 34);
               ^
cc1: some warnings being treated as errors
scripts/Makefile.build:294: recipe for target '/opt/rtlwifi/rtlwifi_new/base.o' failed
make[2]: *** [/opt/rtlwifi/rtlwifi_new/base.o] Error 1
Makefile:1524: recipe for target '_module_/opt/rtlwifi/rtlwifi_new' failed
make[1]: *** [_module_/opt/rtlwifi/rtlwifi_new] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-4.10.0-35-generic'
Makefile:47: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by AryToNeX** [comment]
Built the "kernel_version" branch and I tested it for some hours.

TL;DR: Nothing has changed.

WiFi keeps disconnecting, page loading is slow and I have to reload them often. YouTube video playback is unenjoyable (frame drops, video/audio lag, the spinning circle appearing every 20 seconds). I can't even do speed tests because wifi disconnects when the speed test is running.

**Comment by jhod0** [comment]
Same here as AryToNex, this fixed nothing for me. I'm running kubuntu with kernel 4.10 on an idea pad 320

**Comment by mirh** [comment]
Over on manjaro we had people claiming to have fixed the problem
https://forum.manjaro.org/t/wifi-fails-time-to-time-rtl8821ae/28914/9

**Comment by jhod0** [comment]
This is still very much not working for me, even uninstalling and following the instructions from Manjaro. Wifi drops after about ~5min sometimes, an hour if I'm lucky.

I had created an /etc/modprobe.d/rtl8821ae.conf with some options to try to improve the situation, I removed it when I reinstalled by the Manjaro instructions.

**Comment by christianmtr** [comment]
I had installed Xubuntu 16.04 and the connection falls many times.
I installed Ubuntu Mate 16.04 some weeks ago and the connections work fine all time, no one drops never. I don't know why, I didn't anything. I just installed this repo and works fine.
`uname -r` command shows: `4.10.0-37-generic`

**Comment by mirh** [comment]
I just spotted the supposed fix landing in 4.14-rc6. 

Test it.

**Comment by NickBT** [comment]
Ubuntu mainline updated my kernel to 4.13,0-26, which appears to have fixed the bug which I originally posted. It has taken an enormously long time for this to happen.

**Comment by lwfinger** [comment]
And just how am I to affect the rate of change of Ubuntu kernels????? I submitted the fixes to mainline AS SOON AS THE BUG WAS FOUND. Everything else was out of my control!!!!!

I know - I will just assign future bugs to you since you seem to know what to do.
